src/MDMTransaction.m

```diff
@@ -30,10 +32,25 @@ - (instancetype)init {
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
-  [_logs addObject:log];
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan named:(NSString *)name toTarget:(id)target {
+  NSParameterAssert(name.length > 0);
+  [self commonRemovePlan:plan fromTarget:target named:name];
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  [self commonRemovePlan:nil fromTarget:target named:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  [_logs addObject:[[MDMTransactionLog alloc] initWithPlan:[plan copy] target:target name:name removal:FALSE]];
+}
+
+- (void)commonRemovePlan:(NSObject<MDMPlan> *)plan fromTarget:(nonnull id)target named:(NSString *)named {
+  [_logs addObject:[[MDMTransactionLog alloc] initWithPlan:plan target:target name:named removal:TRUE]];
```

Comment:
Perhaps we should move the `[plan copy]` code into the MDMTransactionLog's initializer? Need to enforce the fact that plans are copied into a transaction the moment they're added.